src/data/set/function.lean

```diff
@@ -418,6 +428,21 @@ lemma surj_on.maps_to_compl (h : surj_on f s t) (h' : injective f) : maps_to f s
 lemma maps_to.surj_on_compl (h : maps_to f s t) (h' : surjective f) : surj_on f sᶜ tᶜ :=
 h'.forall.2 $ λ x ht, mem_image_of_mem _ $ λ hs, ht (h hs)
 
+lemma eq_on.cancel_right (hf : s.eq_on (g₁ ∘ f) (g₂ ∘ f)) (hf' : s.surj_on f t) : t.eq_on g₁ g₂ :=
+begin
+  intros b hb,
+  obtain ⟨a, ha, rfl⟩ := hf' hb,
+  exact hf ha,
+end
+
+lemma surj_on.cancel_right (hf : s.surj_on f t) (hf' : s.maps_to f t) :
+  s.eq_on (g₁ ∘ f) (g₂ ∘ f) ↔ t.eq_on g₁ g₂ :=
+⟨λ h, h.cancel_right hf, λ h, h.comp_right hf'⟩
+
+lemma inj_on.cancel_left (hg : t.inj_on g) (hf₁ : s.maps_to f₁ t) (hf₂ : s.maps_to f₂ t) :
```

Comment:
Surely this doesn't belong under the "surjectivity" heading?

Reply:
Whoops